docker-compose.yml

```diff
@@ -161,21 +161,22 @@ services:
       KAFKA_LDAP_JAVA_NAMING_FACTORY_INITIAL: com.sun.jndi.ldap.LdapCtxFactory
       KAFKA_LDAP_COM_SUN_JNDI_LDAP_READ_TIMEOUT: 3000
       KAFKA_LDAP_JAVA_NAMING_PROVIDER_URL: ldap://openldap:389
-      # How to authenticate to LDAP
+      # Authenticate to LDAP
       KAFKA_LDAP_JAVA_NAMING_SECURITY_PRINCIPAL: cn=admin,dc=confluentdemo,dc=io
       KAFKA_LDAP_JAVA_NAMING_SECURITY_CREDENTIALS: admin
       KAFKA_LDAP_JAVA_NAMING_SECURITY_AUTHENTICATION: simple
-      # How to locate users and groups
-      KAFKA_LDAP_SEARCH_MODE: USERS
+      # Locate LDAP users and groups
+      KAFKA_LDAP_SEARCH_MODE: GROUPS
+      KAFKA_LDAP_GROUP_SEARCH_BASE: ou=groups,dc=confluentdemo,dc=io
+      KAFKA_LDAP_GROUP_NAME_ATTRIBUTE: cn
+      KAFKA_LDAP_GROUP_MEMBER_ATTRIBUTE: memberUid
+      KAFKA_LDAP_GROUP_OBJECT_CLASS: posixGroup
+      KAFKA_LDAP_GROUP_MEMBER_ATTRIBUTE_PATTERN: cn=(.*),ou=users,dc=confluentdemo,dc=io
       KAFKA_LDAP_USER_SEARCH_BASE: ou=users,dc=confluentdemo,dc=io
-      KAFKA_LDAP_GROUP_SEARCH_BASE: ou=users,dc=confluentdemo,dc=io
       KAFKA_LDAP_USER_NAME_ATTRIBUTE: uid
       KAFKA_LDAP_USER_OBJECT_CLASS: inetOrgPerson
       KAFKA_LDAP_USER_MEMBEROF_ATTRIBUTE: ou
```

Comment:
So I _think_ since you are using `ldap.search.mode = GROUPS`, we don't need the `ldap.user.*` settings at all. @jeffhuang26 what do you think about that?

As for `memberOf`, that's a weird LDAP thing. You can tell Kafka to crawl LDAP either by starting from users or from groups by using `ldap.search.mode`.  If you are crawling by groups (default), then Kafka first searches for all groups in LDAP, and then looks for users in those LDAP entries. This works well with almost all LDAP servers.

If, on the other hand, you are crawling LDAP by users (non-default), then Kafka first searches for all users, and then looks for a `memberOf` (or similar) attribute to make a list of all known groups. And that's where the problem arises, because the standardized LDAP schemas only put a pointer from group LDAP records to user LDAP records. To get the reverse pointers (i.e. from LDAP users to LDAP groups), some LDAP servers use a virtual attribute, which is what `memberOf` is. The behavior varies across LDAP servers, but it's a good bet to assume that it is created on the fly, and not directly editable. Specifically, AD uses `memberOf`, OpenLdap doesn't.

For reference, here's what the current LDIF looks like in this demo. I used `docker exec openldap ldapsearch -x -H ldap://localhost -b dc=confluentdemo,dc=io -D "cn=admin,dc=confluentdemo,dc=io" -w admin` for extracting this once the demo was up and running.

```
# users, confluentdemo.io
dn: ou=users,dc=confluentdemo,dc=io
objectClass: organizationalUnit
ou: Users

# groups, confluentdemo.io
dn: ou=groups,dc=confluentdemo,dc=io
objectClass: organizationalUnit
ou: Groups

# Kafka Developers, groups, confluentdemo.io
dn: cn=Kafka Developers,ou=groups,dc=confluentdemo,dc=io
objectClass: top
objectClass: posixGroup
cn: Kafka Developers
gidNumber: 5000
memberUid: cn=alice,ou=users,dc=confluentdemo,dc=io
memberUid: cn=barnie,ou=users,dc=confluentdemo,dc=io

# alice, users, confluentdemo.io
dn: cn=alice,ou=users,dc=confluentdemo,dc=io
objectClass: inetOrgPerson
objectClass: posixAccount
objectClass: shadowAccount
uid: alice
sn: LookingGlass
givenName: Alice
cn: alice
displayName: Alice LookingGlass
uidNumber: 10000
gidNumber: 5000
userPassword:: [removed]
gecos: alice
loginShell: /bin/bash
homeDirectory: /home/alice
```

Reply:
Here's the set of config options to use for locating users and groups (both under `kafka1` and `kafka2`):

```
      # Locate LDAP users and groups
      KAFKA_LDAP_SEARCH_MODE: GROUPS
      KAFKA_LDAP_GROUP_SEARCH_BASE: ou=groups,dc=confluentdemo,dc=io
      KAFKA_LDAP_GROUP_NAME_ATTRIBUTE: cn
      KAFKA_LDAP_GROUP_MEMBER_ATTRIBUTE: memberUid
      KAFKA_LDAP_GROUP_OBJECT_CLASS: posixGroup
      KAFKA_LDAP_GROUP_MEMBER_ATTRIBUTE_PATTERN: cn=(.*),ou=users,dc=confluentdemo,dc=io
      KAFKA_LDAP_USER_SEARCH_BASE: ou=users,dc=confluentdemo,dc=io
      KAFKA_LDAP_USER_NAME_ATTRIBUTE: uid
      KAFKA_LDAP_USER_OBJECT_CLASS: inetOrgPerson
```
`KAFKA_LDAP_USER_MEMBEROF_ATTRIBUTE` and `KAFKA_LDAP_USER_MEMBEROF_ATTRIBUTE_PATTERN` are extraneous (and incorrect for the current LDAP schema).

Reply:
I will test this today.  Agree with the USER settings comments, those are leftovers from when I was trying to get USERS search mode to work.

Reply:
> KAFKA_LDAP_USER_MEMBEROF_ATTRIBUTE and KAFKA_LDAP_USER_MEMBEROF_ATTRIBUTE_PATTERN are extraneous (and incorrect for the current LDAP schema).

In the latest PR commit, I have removed these lines and verified the demo still works.

Reply:
> So I _think_ since you are using `ldap.search.mode = GROUPS`, we don't need the `ldap.user.*` settings at all. @jeffhuang26 what do you think about that?
> 
> As for `memberOf`, that's a weird LDAP thing. You can tell Kafka to crawl LDAP either by starting from users or from groups by using `ldap.search.mode`. If you are crawling by groups (default), then Kafka first searches for all groups in LDAP, and then looks for users in those LDAP entries. This works well with almost all LDAP servers.
> 
> If, on the other hand, you are crawling LDAP by users (non-default), then Kafka first searches for all users, and then looks for a `memberOf` (or similar) attribute to make a list of all known groups. And that's where the problem arises, because the standardized LDAP schemas only put a pointer from group LDAP records to user LDAP records. To get the reverse pointers (i.e. from LDAP users to LDAP groups), some LDAP servers use a virtual attribute, which is what `memberOf` is. The behavior varies across LDAP servers, but it's a good bet to assume that it is created on the fly, and not directly editable. Specifically, AD uses `memberOf`, OpenLdap doesn't.
> 
> For reference, here's what the current LDIF looks like in this demo. I used `docker exec openldap ldapsearch -x -H ldap://localhost -b dc=confluentdemo,dc=io -D "cn=admin,dc=confluentdemo,dc=io" -w admin` for extracting this once the demo was up and running.
> 
> ```
> # users, confluentdemo.io
> dn: ou=users,dc=confluentdemo,dc=io
> objectClass: organizationalUnit
> ou: Users
> 
> # groups, confluentdemo.io
> dn: ou=groups,dc=confluentdemo,dc=io
> objectClass: organizationalUnit
> ou: Groups
> 
> # Kafka Developers, groups, confluentdemo.io
> dn: cn=Kafka Developers,ou=groups,dc=confluentdemo,dc=io
> objectClass: top
> objectClass: posixGroup
> cn: Kafka Developers
> gidNumber: 5000
> memberUid: cn=alice,ou=users,dc=confluentdemo,dc=io
> memberUid: cn=barnie,ou=users,dc=confluentdemo,dc=io
> 
> # alice, users, confluentdemo.io
> dn: cn=alice,ou=users,dc=confluentdemo,dc=io
> objectClass: inetOrgPerson
> objectClass: posixAccount
> objectClass: shadowAccount
> uid: alice
> sn: LookingGlass
> givenName: Alice
> cn: alice
> displayName: Alice LookingGlass
> uidNumber: 10000
> gidNumber: 5000
> userPassword:: [removed]
> gecos: alice
> loginShell: /bin/bash
> homeDirectory: /home/alice
> ```

Good catch!